In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
from tensorflow.keras.callbacks import LearningRateScheduler
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras import layers
from tensorflow.keras import models
from keras.callbacks import ModelCheckpoint
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow import keras


import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report
from sklearn.preprocessing import MinMaxScaler
import optuna
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix

import pandas as pd
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
from tensorflow.keras.callbacks import LearningRateScheduler
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras import layers
from tensorflow.keras import models
from sklearn.metrics import confusion_matrix, matthews_corrcoef
from sklearn.metrics import roc_curve, auc
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt

from sklearn.neighbors import KNeighborsClassifier


from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB

from sklearn.preprocessing import MinMaxScaler

import xgboost as xgb
import lightgbm as lgb
from catboost import CatBoostClassifier

In [2]:
def print_error_matrix(y_true, y_pred):
    cm = confusion_matrix(y_true, y_pred)
    plt.figure(figsize=(10, 8))
    sns.heatmap(cm, annot=True, fmt='g', cmap='coolwarm')
    plt.xlabel('Prediction')
    plt.ylabel('True')
    plt.show()

def plot_ROC(fpr, tpr, roc_auc):
    plt.figure()
    plt.plot(fpr, tpr, color='darkorange', lw=2, 
             label=f'ROC curve (AUC = {roc_auc:.2f})')
    plt.plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--') 
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver Operating Characteristic (ROC)')
    plt.legend(loc="lower right")
    plt.grid(True)
    plt.show()

def specificity_score(y_true, y_pred):
    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
    return tn / (tn + fp) if (tn + fp) > 0 else 0

def evaluate_model(model, scaler="MinMaxScaler",val=0,disp=0):
    # Wczytanie danych
    train = pd.read_csv("dataset_train.csv")
    test = pd.read_csv("dataset_test.csv")

    # Binarization
    train['arrytmia'] = train['arrytmia'].apply(lambda x: 1 if x > 1 else x)
    test['arrytmia'] = test['arrytmia'].apply(lambda x: 1 if x > 1 else x)

    if scaler == "MinMaxScaler":
        scaler_obj = MinMaxScaler()
    else:
        test['arrytmia'] = test['arrytmia'].apply(lambda x: -1 if x == 0 else x)
        train['arrytmia'] = train['arrytmia'].apply(lambda x: -1 if x == 0 else x)
        scaler_obj = StandardScaler()


    X_train = train.iloc[:, 1:]
    y_train = train['arrytmia'].values
    if val==0:
        X_test = test.iloc[:, 1:]
        y_test = test['arrytmia'].values
    if val==1:
        X_train, X_val, y_train, y_val = train_test_split( X_train, y_train, test_size=0.2, random_state=42)
        X_test=X_val
        y_test=y_val

    X_train = train.iloc[:, 1:]
    y_train = train['arrytmia'].values
    X_test = test.iloc[:, 1:]
    y_test = test['arrytmia'].values

    # Skalowanie
    X_train = pd.DataFrame(scaler_obj.fit_transform(X_train), columns=X_train.columns)
    X_test = pd.DataFrame(scaler_obj.transform(X_test), columns=X_test.columns)

    # Trenowanie modelu
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    # Miary
    tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()

    acc = (tp + tn) / (tp + tn + fp + fn) if (tp + tn + fp + fn) > 0 else 0
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0
    specificity = tn / (tn + fp) if (tn + fp) > 0 else 0
    f1 = 2 * tp / (2 * tp + fp + fn) if (2 * tp + fp + fn) > 0 else 0
    mcc = matthews_corrcoef(y_test, y_pred)
    nMCC = (mcc + 1) / 2
    fc = 0.5 * f1 + 0.3 * specificity + 0.2 * nMCC

    if disp:
        print(f"Accuracy:     {acc:.2f}")
        print(f"Specificity:  {specificity:.2f}")
        print(f"F1-score:     {f1:.2f}")
        # print(f"Precision:    {precision:.2f}")
        # print(f"Recall:       {recall:.2f}")
        # print(f"MCC:          {mcc:.2f}")
        print(f"nMCC:         {nMCC:.2f}")
        print(f"Funkcja celu: {fc:.2f}")
    return fc

## KNN

In [3]:
n_trials=100

In [4]:
# def objective(trial):
#     n_neighbors = trial.suggest_int('n_neighbors', 30, 300) # Liczba sąsiadów
#     p = trial.suggest_int('p', 1, 10)           # Liczba sąsiadów
    
#     model = KNeighborsClassifier(
#         n_neighbors=n_neighbors,
#         p=p,
#     )

#     score=evaluate_model(model,val=1)
#     return round(score,4) 

# study = optuna.create_study(direction='maximize')
# study.optimize(objective, n_trials=n_trials)

# best_params = study.best_params
# print("Najlepsze parametry:", best_params)

In [5]:
# best_model = KNeighborsClassifier(**best_params)
# score=evaluate_model(best_model,disp=1)

## DecisionTree

In [6]:
# def objective(trial):
#     criterion = trial.suggest_categorical('criterion', ['gini', 'entropy']) 
#     max_depth = trial.suggest_int('max_depth', 3, 30) 
#     min_samples_split = trial.suggest_int('min_samples_split', 2, 20)
#     min_samples_leaf = trial.suggest_int('min_samples_leaf', 2, 20)
#     max_features = trial.suggest_categorical('max_features', ['sqrt', 'log2', None])

#     model = DecisionTreeClassifier(
#         criterion=criterion,
#         max_depth=max_depth,
#         min_samples_split=min_samples_split,
#         min_samples_leaf=min_samples_leaf,
#         max_features=max_features,
#         random_state=42
#     )


#     score=evaluate_model(model,val=1)
#     return round(score,4)  

# study = optuna.create_study(direction='maximize')
# study.optimize(objective, n_trials=n_trials)

# best_params = study.best_params
# print("Najlepsze parametry:", best_params)

In [7]:
# best_model = DecisionTreeClassifier(**best_params, random_state=42)
# score=evaluate_model(best_model,disp=1)

## RandomForestClassifier

In [8]:
# def objective(trial):
#     n_estimators = trial.suggest_int('n_estimators', 1, 200) # liczba drzew w lesie
#     criterion = trial.suggest_categorical('criterion', ['gini', 'entropy'])
#     max_depth = trial.suggest_int('max_depth', 3, 30) # Maksymalna głebokość drzewa
#     min_samples_split = trial.suggest_int('min_samples_split', 2, 20) # Minimalna liczba próbek w węźle wymagana do podzielenia
#     min_samples_leaf = trial.suggest_int('min_samples_leaf', 2, 20) # minimalna liczba próbek w liściu
#     max_features = trial.suggest_categorical('max_features', ['sqrt', 'log2', None]) # Liczba cech rozważanych podczas rodzielania węzła
    
#     model = RandomForestClassifier(
#         criterion=criterion,
#         n_estimators=n_estimators,
#         max_depth=max_depth,
#         min_samples_split=min_samples_split,
#         min_samples_leaf=min_samples_leaf,
#         max_features=max_features,
#         random_state=42
#     )

    
#     score=score=evaluate_model(model,val=1)
#     return round(score,4)   

# study = optuna.create_study(direction='maximize')
# study.optimize(objective, n_trials=n_trials)

# best_params = study.best_params
# print("Najlepsze parametry:", best_params)


In [9]:
# best_model = RandomForestClassifier(**best_params, random_state=42)
# score=evaluate_model(best_model,disp=1)

## SVM

In [10]:
def objective(trial):
    C = trial.suggest_float('C', 1e-4, 1e4,log=True)
    kernel = trial.suggest_categorical('kernel', ['linear', 'poly', 'rbf', 'sigmoid'])
    
    
    if kernel == "linear":
        model = SVC(C=C, kernel=kernel, random_state=42)
    
    gamma = trial.suggest_categorical('gamma', ['scale', 'auto'])
    
    if kernel == 'poly':
        degree = trial.suggest_int('degree', 2, 4)
        coef0 = trial.suggest_float ('coef0', -1, 1)
        model = SVC(C=C, kernel=kernel, gamma=gamma, degree=degree, coef0=coef0, random_state=42)
    if kernel == 'sigmoid':
        coef0 = trial.suggest_float ('coef0', -1, 1)
        model = SVC(C=C, kernel=kernel, gamma=gamma, coef0=coef0, random_state=42)
    else:
        model = SVC(C=C, kernel=kernel, gamma=gamma, random_state=42)
    
    
    score=evaluate_model(model,val=1)
    return round(score,4)  

study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=n_trials)

best_params = study.best_params
print("Najlepsze parametry:", best_params)


[I 2025-08-11 22:41:03,385] A new study created in memory with name: no-name-9864e40f-4e8f-422a-958a-34006e60aa7e
[I 2025-08-11 22:42:41,042] Trial 0 finished with value: 0.6177 and parameters: {'C': 0.07650230939282066, 'kernel': 'linear', 'gamma': 'auto'}. Best is trial 0 with value: 0.6177.
[I 2025-08-11 22:44:46,052] Trial 1 finished with value: 0.4519 and parameters: {'C': 0.0021212655804023046, 'kernel': 'rbf', 'gamma': 'scale'}. Best is trial 0 with value: 0.6177.
[I 2025-08-11 22:46:14,655] Trial 2 finished with value: 0.4805 and parameters: {'C': 0.7875914231457218, 'kernel': 'sigmoid', 'gamma': 'scale', 'coef0': -0.6930517708885819}. Best is trial 0 with value: 0.6177.
[I 2025-08-11 22:48:52,509] Trial 3 finished with value: 0.7221 and parameters: {'C': 6.562690210120786, 'kernel': 'linear', 'gamma': 'auto'}. Best is trial 3 with value: 0.7221.
[I 2025-08-11 22:51:36,810] Trial 4 finished with value: 0.7607 and parameters: {'C': 31.553147622410403, 'kernel': 'poly', 'gamma': 

Najlepsze parametry: {'C': 63.14720694278472, 'kernel': 'sigmoid', 'gamma': 'auto', 'coef0': -0.5689847583962766}


[I 2025-08-12 00:14:05,955] Trial 37 finished with value: 0.7874 and parameters: {'C': 63.14720694278472, 'kernel': 'sigmoid', 'gamma': 'auto', 'coef0': -0.5689847583962766}. Best is trial 37 with value: 0.7874.

In [12]:
best_model = SVC(**best_params, random_state=42)
score=evaluate_model(best_model,disp=1)

Accuracy:     0.80
Specificity:  0.72
F1-score:     0.82
nMCC:         0.80
Funkcja celu: 0.79
